src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,431 @@
+
+	MyDistributedTransactionBackend->databaseId = MyDatabaseId;
+
+	MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier =
+		PG_GETARG_INT64(0);
+	MyDistributedTransactionBackend->transactionId.transactionId = PG_GETARG_INT64(1);
+	MyDistributedTransactionBackend->transactionId.timestamp = PG_GETARG_TIMESTAMPTZ(2);
+
+	SpinLockRelease(&MyDistributedTransactionBackend->mutex);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * get_distributed_transaction_id returns a tuple with (databaseId, initiatorNodeIdentifier,
+ * transactionId, timestamp) that exists in the shared memory associated with this backend.
+ */
+Datum
+get_distributed_transaction_id(PG_FUNCTION_ARGS)
+{
+	ReturnSetInfo *returnSetInfo = (ReturnSetInfo *) fcinfo->resultinfo;
+	TupleDesc tupleDescriptor = NULL;
+	Tuplestorestate *tupleStore = NULL;
+	MemoryContext perQueryContext = NULL;
+	MemoryContext oldContext = NULL;
+
+	int attributeCount = 4;
+	Datum values[attributeCount];
+	bool isNulls[attributeCount];
+
+	DistributedTransactionId *distributedTransctionId = NULL;
+
+	CheckCitusVersion(ERROR);
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (returnSetInfo == NULL || !IsA(returnSetInfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context " \
+						"that cannot accept a set")));
+	}
+
+	if (!(returnSetInfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not " \
+						"allowed in this context")));
+	}
+
+	/* build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupleDescriptor) != TYPEFUNC_COMPOSITE)
+	{
+		elog(ERROR, "return type must be a row type");
+	}
+
+	perQueryContext = returnSetInfo->econtext->ecxt_per_query_memory;
+
+	oldContext = MemoryContextSwitchTo(perQueryContext);
+
+	tupleStore = tuplestore_begin_heap(true, false, work_mem);
+	returnSetInfo->returnMode = SFRM_Materialize;
+	returnSetInfo->setResult = tupleStore;
+	returnSetInfo->setDesc = tupleDescriptor;
+
+	MemoryContextSwitchTo(oldContext);
+
+	/* MyDistributedTransactionBackend should always be avaliable, just out of paranoia */
+	if (!MyDistributedTransactionBackend)
+	{
+		ereport(ERROR, (errmsg("backend is not ready for distributed transactions")));
+	}
+
+	distributedTransctionId = GetCurrentDistributedTransctionId();
+
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	values[0] = ObjectIdGetDatum(MyDatabaseId);
+	values[1] = UInt64GetDatum(distributedTransctionId->initiatorNodeIdentifier);
+	values[2] = UInt64GetDatum(distributedTransctionId->transactionId);
+	values[3] = TimestampTzGetDatum(distributedTransctionId->timestamp);
+
+	tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
```

Comment:
I think you should just `heap_form_tuple()` with the result tupdesc and return the result.